File: AgentLens/Services/Memory/MemoryRecallBudget.swift
import Foundation
import OpenBurnBarCore

/// Maps the "High-recall (per reply)" toggle to concrete `MemoryRecallRequest`
/// parameters, and exposes the wrapper-envelope overhead the recall packer must charge.
///
/// High-recall's lever is the LIMIT — how many distinct approved facts are eligible to
/// surface — NOT a larger token slice. The prompt-token arbiter (`PromptTokenArbiter`)
/// caps the assembled, wrapped `.memory` section at its own `memoryBudget` regardless of
/// what recall asks for, so inflating `tokenBudget` here only forced the arbiter to
/// truncate a wrapped block — which severed the `</UNTRUSTED_CONTENT>` seal (M1/M2 audit
/// findings). The token budget therefore tracks the arbiter's allocation in both modes,
/// and the per-snippet wrapper envelope is charged in `recallChatMemorySnippets` so the
/// section fits the cap without truncation.
enum MemoryRecallBudget {
    /// Default snippet limit used when high-recall is off.
    /// Must match `MemoryRecallRequest.init` default parameter.
    static let defaultLimit: Int = 8

    /// Snippet limit when high-recall is on: a larger candidate pool under the SAME token
    /// cap (more distinct small facts can surface; the arbiter still bounds total size).
    static let highRecallLimit: Int = 16

    /// Per-snippet token cost of the `LLMSafeContent.wrapUntrusted` envelope (open tag +
    /// provenance + close tag + the multi-sentence CRITICAL RULE). The recall packer adds
    /// this to each snippet's body estimate so the budget reflects the WRAPPED size the
    /// arbiter actually sees — otherwise the assembled `.memory` section overflows the
    /// arbiter cap and is truncated. Estimated in the SAME units the `PromptTokenArbiter`
    /// uses for the `.memory` section (prose, ~3.5 chars/token) so recall and the arbiter
    /// agree on cost; using chars/4 here would under-count and re-introduce overflow.
    /// Derived from the real wrapper of an empty body so it tracks the template.
    static let wrapperTokenOverhead: Int = {
        // Price a WORST-CASE provenance: the real one assembled in
        // ChatSessionController+Search.recallMemorySection is "memory:<memoryID>@<jumpID>",
        // where memoryID is a UUID / "memory-<uuid>-<n>" and jumpID can be a message UUID
        // or a "v1-local:<sha256-hex>" cross-device tag — up to ~130 chars, far longer than
        // a short placeholder. Over-pricing the provenance (plus +1 token for the per-snippet
        // "\n\n" join) keeps the recall budget CONSERVATIVE vs the actual wrapped string, so
        // the assembled .memory section cannot overflow the arbiter cap and get truncated.
        let worstCaseProvenance = "memory:" + String(repeating: "x", count: 140)
        let envelope = LLMSafeContent.wrapUntrusted("", provenance: worstCaseProvenance)
        return PromptTokenArbiter.estimateProseTokens(envelope) + 1
    }()

    /// Returns `(limit, tokenBudget)` for a `MemoryRecallRequest`, given the arbiter's
    /// allocated budget and the toggle state.
    ///
    /// - Parameters:
    ///   - arbiterBudget: Token count from `PromptTokenArbiter.memoryBudget` (the cap the
    ///     assembled wrapped section must fit).
    ///   - highRecall: Value of `settingsManager.memoryHighRecallPerReply`.
    /// - Returns: A tuple suitable for `MemoryRecallRequest(query:scope:tokenBudget:limit:)`.
    static func forReply(arbiterBudget: Int, highRecall: Bool) -> (limit: Int, tokenBudget: Int) {
        (limit: highRecall ? highRecallLimit : defaultLimit, tokenBudget: max(arbiterBudget, 1))
    }
}

File: windows/app/OpenBurnBar.App.MemorySearch/Memory/MemoryRecallBudget.cs
using System;
using System.Globalization;

namespace OpenBurnBar.App.MemorySearch.Memory;

// PORTED (faithful) from
//   AgentLens/Services/Memory/MemoryRecallBudget.swift
//   AgentLens/Services/ContextBuilder.swift          (PromptTokenArbiter.estimateProseTokens)
//   OpenBurnBarCore/.../LogParser/TokenExtractionUtility.swift (estimatedTokenCount)
//
// High-recall's lever is the LIMIT (how many distinct approved facts are eligible), NOT a larger
// token slice — the token budget tracks the arbiter's allocation in both modes.

/// <summary>Result of <see cref="MemoryRecallBudget.ForReply"/>. Peer of the Swift tuple.</summary>
public readonly record struct MemoryRecallBudgetResult(int Limit, int TokenBudget);

/// <summary>Maps the high-recall toggle to recall-request parameters. Swift: <c>enum MemoryRecallBudget</c>.</summary>
public static class MemoryRecallBudget
{
    /// <summary>Snippet limit when high-recall is off. Swift: <c>defaultLimit = 8</c>. Must match
    /// the <c>MemoryRecallRequest</c> default.</summary>
    public const int DefaultLimit = 8;

    /// <summary>Snippet limit when high-recall is on. Swift: <c>highRecallLimit = 16</c>.</summary>
    public const int HighRecallLimit = 16;

    /// <summary>Prose token ratio (~3.5 chars/token). Swift: <c>PromptTokenArbiter</c> prose ratio.</summary>
    public const double ProseCharsPerToken = 3.5;

    /// <summary>
    /// Per-snippet token cost of the untrusted-content wrapper envelope. Swift:
    /// <c>wrapperTokenOverhead</c> (a computed 202). Pinned here because the wrapper template
    /// (<c>LLMSafeContent.wrapUntrusted</c> + <c>criticalRule</c>) lives in the Core SharedModels
    /// safety utility, which is outside this memory-search core. Derivation:
    ///   worst-case provenance = "memory:" + 140×'x' (147 chars)
    ///   → the wrapUntrusted envelope of an empty body is 701 chars
    ///   → EstimateProseTokens(701) = ceil(701 / 3.5) = 201
    ///   → + 1 (the per-snippet "\n\n" join) = 202.
    /// A test pins <c>EstimateProseTokens(701) + 1 == 202</c>. If the Core template changes, this
    /// must be re-derived (a bucket-B wiring concern, since the wrap itself is not ported here).
    /// </summary>
    public const int WrapperTokenOverhead = 202;

    /// <summary>
    /// Returns <c>(limit, tokenBudget)</c> for a recall request. Swift:
    /// <c>forReply(arbiterBudget:highRecall:)</c>. <c>limit = highRecall ? 16 : 8</c>;
    /// <c>tokenBudget = max(arbiterBudget, 1)</c>.
    /// </summary>
    public static MemoryRecallBudgetResult ForReply(int arbiterBudget, bool highRecall)
    {
        int limit = highRecall ? HighRecallLimit : DefaultLimit;
        int tokenBudget = Math.Max(arbiterBudget, 1);
        return new MemoryRecallBudgetResult(limit, tokenBudget);
    }

    /// <summary>
    /// Estimated token count for a character count. Swift:
    /// <c>TokenExtractionUtility.estimatedTokenCount(for:charsPerToken:)</c>:
    /// <c>characters &lt;= 0 ? 0 : ceil(characters / charsPerToken)</c>.
    /// </summary>
    public static int EstimatedTokenCount(int characters, double charsPerToken)
    {
        if (characters <= 0)
        {
            return 0;
        }

        return (int)Math.Ceiling(characters / charsPerToken);
    }

    /// <summary>Estimated prose tokens for content (grapheme-count / 3.5, rounded up). Swift:
    /// <c>PromptTokenArbiter.estimateProseTokens(_:)</c>.</summary>
    public static int EstimateProseTokens(string content)
    {
        ArgumentNullException.ThrowIfNull(content);
        int characters = GraphemeCount(content);
        return EstimatedTokenCount(characters, ProseCharsPerToken);
    }

    private static int GraphemeCount(string value)
    {
        int count = 0;
        var enumerator = StringInfo.GetTextElementEnumerator(value);
        while (enumerator.MoveNext())
        {
            count++;
        }

        return count;
    }
}
